ui/kotlinx-coroutines-android/src/HandlerDispatcher.kt

```diff
@@ -185,22 +185,27 @@ private var choreographer: Choreographer? = null
 public suspend fun awaitFrame(): Long {
     // fast path when choreographer is already known
     val choreographer = choreographer
-    if (choreographer != null) {
-        return suspendCancellableCoroutine { cont ->
+    return if (choreographer != null) {
+        suspendCancellableCoroutine { cont ->
             postFrameCallback(choreographer, cont)
         }
+    } else {
+        awaitFrameSlowPath()
     }
-    // post into looper thread to figure it out
-    return suspendCancellableCoroutine { cont ->
-        Dispatchers.Main.dispatch(EmptyCoroutineContext, Runnable {
+}
+
+private suspend fun awaitFrameSlowPath(): Long = suspendCancellableCoroutine { cont ->
+    if (Looper.myLooper() === Looper.getMainLooper()) { // Check if we are already in the main looper thread
+        updateChoreographerAndPostFrameCallback(cont)
```

Comment:
I saw a discussion about `Dispatchers.Main.immediate` not being good enough for this and something about intending to fix it. Could you elaborate?

Reply:
Well, I was hoping that adding `Dispatchers.main.immediate` would solve the problem, but it won't due to how we do structure our code: `CancellableContinuation` already incapsulates the dispatcher (for this case -- `Dispatchers.Main`) and we cannot "just" resume it in place.

Neither can we do `Dispatchers.Main.immediate.resumeUndispatched(cont)` due to how it is implemented -- that's a grey area in "how should 'resumeUndispatched' behave if extension receiver and actual dispatcher do not match". And they do not match because `Dispatchers.Main.immediate` and `Dispatchers.Main` are different dispatchers, non-transparent to each other